### PR TITLE
446 improve res mod perf

### PR DESCRIPTION
**Pull request by hyi**
This branch is slated for NFIE hotfix release, and deployed on park for testing. Please review the code at your earliest convenience so that it can be merged into develop for the NFIE hotfix. It mainly resolves the performance issue with big resource metadata editing by doing on-demand bagging. Another two issues below are also resolved in this branch:

1. Hide the ajax spin wheel icon when irods browser gets invoked again - a minor bug fix.
2. bagit rule fix that stores generated bagit files such as bagit.txt and manifest files to HydroShare default resource rather than the default iRODS demoResc resource.

### CODE REVIEW

File: hs_core/hydroshare/hs_bagit.py [hunk 1]
```diff
@@ -21,20 +22,14 @@ def delete_bag(resource):
...
-    execution of an ibun command to zip up the bagit file hierarchy.
-
-    Note, this procedure may take awhile.  It is highly advised that it be deferred to a Celery task.
-
-    Parameters:
-    :param resource: (subclass of AbstractResource) A resource to create a bag for.
+    # delete the bags table
+    for bag in resource.bags.all():
+        try:
+            bag.delete()
```
Comment (selimnairb) [inline hs_core/hydroshare/hs_bagit.py]: remove whitespace between *try:* and *bag.delete()*.

File: hs_core/hydroshare/hs_bagit.py [hunk 2]
```diff
@@ -21,20 +22,14 @@ def delete_bag(resource):
...
+        except Exception as ex:
+            raise Exception(ex.message)
```
Comment (selimnairb) [inline hs_core/hydroshare/hs_bagit.py]: Remove whitespace before *raise ...*

File: hs_core/hydroshare/hs_bagit.py [hunk 3]
```diff
@@ -12,6 +12,7 @@
 from django_irods.storage import IrodsStorage
 
 def delete_bag(resource):
+
     res_id = resource.short_id
```
Comment (selimnairb) [inline hs_core/hydroshare/hs_bagit.py]: remove whitespace before *res_id...*

File: hs_core/hydroshare/hs_bagit.py [hunk 4]
```diff
@@ -21,20 +22,14 @@ def delete_bag(resource):
...
+    # delete the bags table
+    for bag in resource.bags.all():
+        try:
+            bag.delete()
+        except Exception as ex:
+            raise Exception(ex.message)
 
-    :return: the hs_core.models.Bags instance associated with the new bag.
-    """
+def create_bag_files(resource):
```
Comment (selimnairb) [inline hs_core/hydroshare/hs_bagit.py]: need docstrings documenting this function.

File: hs_core/hydroshare/hs_bagit.py [hunk 5]
```diff
@@ -130,16 +125,43 @@ def create_bag(resource):
     to_file_name = os.path.join(resource.short_id, 'data', 'resourcemap.xml')
     istorage.saveFile(from_file_name, to_file_name, False)
```
Comment (selimnairb) [inline hs_core/hydroshare/hs_bagit.py]: How do we know whether istorage.saveFile() succeeded or failed?

File: hydroshare/irods/ruleGenerateBagIt_HS.r [hunk 6]
```diff
@@ -74,5 +74,5 @@ generateBagIt {
   ### - writes to rodsLog
   msiWriteRodsLog("BagIt bag files created in place: *BAGITDATA <- *BAGITDATA",*Status);
 }
-INPUT *BAGITDATA="/dummy/dummy/dummy" 
+INPUT *BAGITDATA="/hydrotest41Zone/home/hyi", *DESTRESC="demoResc"
```
Comment (selimnairb) [inline hydroshare/irods/ruleGenerateBagIt_HS.r]: This looks like it is specific to your development environment.

File: hs_core/hydroshare/hs_bagit.py [hunk 7]
```diff
@@ -21,20 +22,14 @@ def delete_bag(resource):
...
+            bag.delete()
+        except Exception as ex:
+            raise Exception(ex.message)
 
-    :return: the hs_core.models.Bags instance associated with the new bag.
-    """
+def create_bag_files(resource):
     from . import utils as hs_core_utils
     DATE_FORMAT = "YYYY-MM-DDThh:mm:ssTZD"
     istorage=IrodsStorage()
```
Comment (selimnairb) [inline hs_core/hydroshare/hs_bagit.py]: Regarding:

    dest_prefix = getattr(settings, 'BAGIT_TEMP_LOCATION', '/tmp/hydroshare/')

What happens if the directory represented by dest_prefix does not exist?

File: hydroshare/irods/ruleGenerateBagIt_HS.r (hunk 6 above)
Comment (hyi) [inline hydroshare/irods/ruleGenerateBagIt_HS.r]: This is just a dummy placeholder, and will be overriden by parameters being passed in at run time, but I will change it to be "dummy" to make it clear.

File: hs_core/hydroshare/hs_bagit.py (hunk 5 above)
Comment (hyi) [inline hs_core/hydroshare/hs_bagit.py]: saveFile() runs an irods iput session which will raise SessionException if it fails.

File: hs_core/hydroshare/hs_bagit.py (hunk 7 above)
Comment (hyi) [inline hs_core/hydroshare/hs_bagit.py]: the directory will be created if it does not exist.

**Comment by hyi** [comment]
I have updated the code to use iRODS metadata AVU pairs to keep track of whether a bag is modified for a resource and also used iRODS metadata AVU to lock the bag while the bag is being updated to avoid potential race condition that could result from multiple requests modifying the same bag. Also changed to not create the bag initially when the resource is created, but rather to set the bag_modified AVU metadata to True for the resource so that the bag will be created when it is downloaded. Tested all scenarios including the potential race condition and all seem to work as expected. Will ask Michael to update the code on park.hydroshare.org for team to test. In the meanwhile, please review the code at your earliest convenience so that it can be merged for NFIE hotfix release.

File: hs_core/hydroshare/hs_bagit.py (hunk 2 above)
Comment (pkdash) [inline hs_core/hydroshare/hs_bagit.py]: Is the use of ```try ... except``` here to express the intent that bag.delete() may fail?

Comment (ironfroggy) [inline hs_core/hydroshare/hs_bagit.py]: This `try/except` does nothing but obscure the traceback.

Comment (hyi) [inline hs_core/hydroshare/hs_bagit.py]: @pkdash @ironfroggy The previous code used try/except for bag.delete, but rather than raise an exception, it just ignores/suppresses the error by using "pass" in except: block. I thought it is better to raise the error, but now that I think about it more, the error is raised anyway, so I should not add another layer. So I have removed try/except block here and pushed the code. If any of you see we should ignore the exception here as we did before, let me know as I think we should raise the exception here rather than ignore it.

**Comment by selimnairb** [comment]
@hyi I tested the REST API against park.hydroshare.org using hs_restclient.  Resource creation, making public, and deletion all work as expected.

@valentinedwv Are we ready to do any load testing using hs_restclient?  This would be useful to test for race conditions in the resource bag AVU locking and spin lock code that @hyi wrote.

**Comment by hyi** [comment]
Have implemented everything I can do for the on-demand bagging for the NFIE hotfix (which most likely needs to be deployed this coming Monday) and ask for final code review. The latest changes I made include:
1. used irods metadata AVU to keep track of whether a resource is dirty with its bag to be re-created when downloading.
2. removed AJAX download to revert it back to the previous url/view mapping for download, hence spin wheel progress indicator does not show any more.
3. added more check and exception handling for potential race conditions that could result from clashes between multiple irods processes triggered from multiple requests, e.g., when user A is downloading a resource bag that triggers bag-creation on iRODS side, user B deletes this same resource or modifies metadata of this same resource, and initiates a download also at the same time. This is only workaround implemented to avoid exceptions raised due to these potential race conditions. In the future, we need to use external services/packages to ensure process synchronization as recommended by iRODS expert Hao Xu.

**Comment by hyi** [comment]
The latest code update has been deployed on park.hydroshare.org by Michael for team testing.

**Comment by dtarb** [comment]
@hyi thanks for your work on this.  Given that you posted this on Friday, lets not rush a Monday deploy.  Lets review it properly and deploy after review and testing.

File: hs_core/templates/pages/genericresource.html [hunk 8]
```diff
@@ -1416,6 +1416,9 @@ <h4 class="modal-title">Confirm file deletion</h4>
             </div>
             </div>
         <!-- ==== Tabs end ==== -->
+        <div class="rating">
```
Comment (selimnairb) [inline hs_core/templates/pages/genericresource.html]: how is this related to the hotfix?

File: irods_browser_app/static/js/irods.js [hunk 9]
```diff
@@ -34,6 +34,7 @@ $('#btn-signin-irods').on('click',function(){
 $('#btn-select-irods-file').on('click',function(){
     $('#res_type').val($('#resource-type').val());
     $('#file_struct').children().remove();
+    $('.ajax-loader').hide();
```
Comment (selimnairb) [inline irods_browser_app/static/js/irods.js]: This appears to be leftover from when you were using the AJAX loader.

**Comment by selimnairb** [comment]
Pending @ironfroggy's review, I think this is ready to merge.  (though this is not the last word on the subject of downloads from iRODS, though I think @hyi has done a good job to improve metadata editing performance.)

File: irods_browser_app/static/js/irods.js (hunk 9 above)
Comment (hyi) [inline irods_browser_app/static/js/irods.js]: @selimnairb No, this is not a left-over code, but instead it is a minor fix I slated in for iRODS browser I put in which also shows the same spin-wheel when retrieving collection and data objects from iRODS for browsing purposes. I found I did not hide it properly which will have spin-wheel being displayed from start to finish if the user click browse button again. This one line addition will fix that issue.

File: hs_core/templates/pages/genericresource.html (hunk 8 above)
Comment (hyi) [inline hs_core/templates/pages/genericresource.html]: This is another fix I slated in for a regression error that was probably introduced by some conflict in merging along the way (this issue was raised in github by @Castronova). Specifically, the rating for resource code was removed (which was present on beta, but not on www), so I am adding the code back so that users can rate resources as well as rate comments to resolve that regression error. My branch commit history should give more details on this.
